### PR TITLE
add more decorated info when logging error messages

### PR DESCRIPTION
**Pull request by AntonioLok**
# Description

This PR implements the follow:
1- Add more information to the error log in order to help developers with debugging issues when the application throws an error.

# Test
1 - Install dependencies using `npm i`.
2 - Start the app `npm run start`.
3 - 
In a separate terminal, run:
` curl http://localhost:8000/carts`

Now, in the console, you should see the following:
```
2019-10-03T05:06:09.351Z info: GET /carts
error jwt must be provided getDecodedToken C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\utils\middlewares\get-decoded-token.js 21
```

To do another test, run in the separate terminal:
`curl -X POST http://localhost:8000/users/login`

Then , in the console you should see:
```
2019-10-03T05:06:24.598Z info: POST /users/login
error Error: Illegal arguments: undefined, undefined router.post C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\views\routes\users.js 
53
```

To do another test, run in the separate terminal:
`curl http://localhost:8000/products/213`

Then , in the console you should see:
```
2019-10-03T05:06:32.169Z info: GET /products/213
error Cast to ObjectId failed for value "213" at path "_id" for model "Product" router.get C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\views\routes\products.js 24
```
_______________________________
Additional test:

Add the following code  in the file `online-shopping-reactJS-Backend/src/index.js`
First: import the logger function:
`const logger = require('./utils/logger');`

and then add:
```
logger.info('Info message');
logger.info(['Info message']);
logger.warn('Warn message');
logger.warn(['Warn message']);
logger.error('Error message');
logger.error(Error('Error message'));
```
Restart the app. Now in the console, we should see:
```
info Info message anonymous C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\index.js 25
info Info message anonymous C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\index.js 26
warn Warn message anonymous C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\index.js 27
warn Warn message anonymous C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\index.js 28
error Error message anonymous C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\index.js 29
error Error message anonymous C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\index.js 30
```

### CODE REVIEW

**Comment by AntonioLok** [comment]
> This is a good start. The message printed in the console displays the error level 👍 , and the message 👍 . However, it also includes the entire stacktrace. And, based on where the error was triggered from, I'm only seeing the stack trace that starts in the library file (in node_modules). This doesn't help me to figure out which function in your application triggered the message. Look into that and the callsite module to see grab info on exactly where in your application threw the error.

Sorry for the late reply. I was busy the entire week preparing for my test. I can start working on this now that I am free. Btw... my test went well.

Thank you for the review. I have updated the code to use the callsite module to tell us the filename and function that is throwing the error. I have also updated the PR description with the updated "expected log" that you should receive when you run the sample commands. Hopefully I didn't misunderstand what you were asking for.

File: src/utils/logger-custom-format.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      return `${level}: ${site.getFunctionName() || 'anonymous'} ${site.getFileName()} ${site.getLineNumber()}`;
+    }
+
+    const { req } = message;
+    const { url, method } = req;
+    const currentTimestamp = (new Date()).toISOString();
+    return `${currentTimestamp} ${level}: ${method} ${url}`;
+  });
+};
+
```
Comment (howardhenry) [inline src/utils/logger-custom-format.js]: This is a good start. However, the implementation is tightly coupled with the `express` request object. What happens when you want to log an error function in your application that is not a route controller and does not have a request object?

Comment (AntonioLok) [reply src/utils/logger-custom-format.js]: The error logger is not tightly coupled with the express request object. In fact, it does not rely on the request object. 

I think you are getting confused by these lines line 16-19 in `logger-custom-format.js`:
```
    const { req } = message;
    const { url, method } = req;
    const currentTimestamp = (new Date()).toISOString();
    return `${currentTimestamp} ${level}: ${method} ${url}`;
```

They are being used by the the middleware logRequest in `log-request.js` which I used to print info about the incoming request.

Comment (howardhenry) [reply src/utils/logger-custom-format.js]: Try using the error-logger outside of a route controller and you'll see the tight coupling. For example:
1. go to _src/index.js_
2. import the logger (`const errorLogger = request('./utils/error-logger');`)
3. log an error anywhere in that file
```
try {
  invalid-javascript
} catch(e) {
  errorLogger.error(e);
}
```

Comment (AntonioLok) [reply src/utils/logger-custom-format.js]: I tried using:

```
try {
  throw new Error('Error thrown from index.js');
} catch (e) {
  errorLogger.error(e);
}
```

and when I start the app, I get this:
```
> shopping-site-reactjs-backend@1.0.0 start C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend
> node src

error: anonymous C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\index.js 43
Running on localhost:8000
We are connected
(node:22936) DeprecationWarning: collection.ensureIndex is deprecated. Use createIndexes instead.
```

Comment (howardhenry) [reply src/utils/logger-custom-format.js]: I'm sorry, my feedback was poor. The example that I used to test the code wasn't the example that I shared in my comment. What I tested the code with was:

```
errLogger.error('Something when wrong');
```
which threw an error:
```
~/online-shopping-reactJS-Backend/node_modules/winston/lib/winston/logger.js:307
      throw ex;
      ^

TypeError: Cannot destructure property `url` of 'undefined' or 'null'.
```

When using the error logger, I assumed that as an engineer I can use it to replace `console.log`. But it seems that your code only expects an error object or an express `request` object, which is why I mentioned that your code is tightly coupled to with express.

Given this finding, what are your thoughts on how this should be?

Comment (AntonioLok) [reply src/utils/logger-custom-format.js]: Ah okay, that makes more sense now. Thanks for clarifying, now I get why you said that my error logger utility is somewhat dependent on the request object. You have probably noticed this already but this happens because I "expected" only error objects to be passed to the error logger function. Your example passes a non-error object and so my if statement `    if (info instanceof Error) {` is not triggered and it executes the following lines which relies on the req object. "One" solution I could think was to separate the logic into 2 different format functions. 1 used by the error logger function (which now does not depend on an error object passed) and the other one is used by the middleware which is tightly coupled to the req object.

Comment (howardhenry) [reply src/utils/logger-custom-format.js]: Right. But you only looking at use cases that present themselves to you. You've created a logger that either takes an error object or an express request object.

Without over-engineering a solution, I would encourage you to step back and think about all the **reasonable** use cases that you may need in the _immediate_ or _near future_. Otherwise, you will end up refactoring your work too often, which is not so efficient. I'm finding that you're trying to solve _only_ the problem that is right in front of you. It is sometimes better to start slow in the beginning so you can move fast in the end.

Comment (AntonioLok) [reply src/utils/logger-custom-format.js]: "You've created a logger that either takes an error object or an express request object."

Actually.... I have created a logger which takes anything: error objects, arrays, strings, etc. I am not sure why you would say "either" an error object or request object. I remember testing my code with your example which is a string being passed to the logger:
`
logger.error('Something when wrong'); // A string
`
and that worked fine. I have also tested it with arrays.

_________________________________________

"and think about all the reasonable use cases that you may need in the immediate or near future." 

I think I know which "use cases" you mean. I decided before to not include those use cases since they are not being used anywhere right now. I am not a huge fan of adding functionality or code which is not being used at the current moment. But you are right about "you will end up refactoring your work too often". I have refactored a lot of code and implementation due to this. I guess this is something that I need to work on more...
Nevertheless, I have modified the log message to include the message being passed as well as making it look a bit nicer. I have also added support to log messages of level info, warn and error. I have updated the PR to provide some examples on how to test this and sample outputs.

File: src/utils/logger.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+const logger = level => createLogger({
+  level,
+  transports: new transports.Console(),
+  format: customFormatLoggerUtility(),
+});
+
+exports.info = msg => logger('info').info(msg);
+exports.error = msg => logger('error').error(msg);
+exports.warn = msg => logger('warn').warn(msg);
```
Comment (howardhenry) [inline src/utils/logger.js]: It is redundant to have the log level passed in as a string and then called as function of the logger. In your implementation you're passing in 'info' then calling the `info` method on the logger; same with **error** and **warn**. Was there any particular reason you implemented it this way? You can review the Winston documentation to see how you can use the library's API to have a cleaner implementation of this.

File: src/utils/logger-custom-format.js (hunk 1 above)
Comment (howardhenry) [reply src/utils/logger-custom-format.js]: I agree you don't want to over-engineer your solution. So, it's definitely a balance to decide "will I need this functionality in the near future". I think it is indeed reasonable to assume that you will need that functionality in the near future 👍

Comment (AntonioLok) [reply src/utils/logger-custom-format.js]: Haha yup. Thanks a lot for the review, I have said this and I will say it again, I really appreciate your time and help ^-^

File: src/utils/logger.js (hunk 2 above)
Comment (AntonioLok) [reply src/utils/logger.js]: Yes, there was a particular reason. If we look at line 4-8 of the same file `src/utils/logger.js`:

```
const logger = level => createLogger({
  level,
  transports: new transports.Console(),
  format: customFormatLoggerUtility(),
});
```
We can see that we are passing the type of level to the function `createLogger()`. This is required if we want to use logger.error(), logger.warn() and logger.info() respectively. I mean, if i want to use logger.error(), I need to create a new logger with level 'error' and if I want to use logger.warn(), I need to create a new logger with level 'warn". 

TLDR: In order for me to create a logger, I need to pass in the level type. I am passing 3 different levels (error, warn and info) thus creating 3 different type of loggers. Hope this was clear. I have tried other methods and this was the only one that worked for me

Comment (howardhenry) [reply src/utils/logger.js]: There are a number of log levels in Winston. And, setting the log level in the `createLogger` function sets the **maximum** log level that will be captured when you call the logger (https://github.com/winstonjs/winston/blob/master/README.md#using-logging-levels). And, with the following log level mappings:
```
{ 
  error: 0, 
  warn: 1, 
  info: 2, 
  verbose: 3, 
  debug: 4, 
  silly: 5 
}
```
if you set the log level to `info`, then calling `logger.error`, `logger.warn` and `logger.info` will capture and print the log message. Calling `logger.verbose`, `logger.debug` or `logger.silly` will not be captured or printed.

What you are doing is:
- setting the log level to `error`, then calling `error`
- setting the log level to `warn`, then calling `warn`
- setting the log level to `info`, then calling `info`

So, this makes your implementation redundant as setting your log level to `info` when you create the logger allows you to call `error`, `warn` AND `info`.

I'll let you work out the rest of the solution.

The important bit is to dig deeper. All this included in the Winston documentation 👍

Comment (AntonioLok) [reply src/utils/logger.js]: OMG. Scratch what I said about "if i want to use logger.error(), I need to create a new logger with level 'error' and if I want to use logger.warn(), I need to create a new logger with level 'warn'". That actually was not true.... there was a reason why I thought setting the level was needed in order to use logger.error, logger.warn, etc. I tried that before (creating a logger without specifying the level) and it would always throw an error.... it turned out there was a small bug in my code with the logger.... I have updated the logger function to make it less redundant and got rid of that bug.
